docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,383 @@
+
+## Proposal
+
+The chosen approach was heavily inspired by metal3's ip-address-manager, which in turn seems to have been inspired by Kubernetes' PersistentVolumes. The proposal is twofold and consists of the API contract that should be added to CAPI and two examples to better illustrate how the contract is used by IPAM and infrastructure providers.
+
+### IPAM API Contract
+
+The IPAM API contract provides a generic interface between CAPI infrastructure providers and to-be-developed IPAM providers. Any interested infrastructure is supposed to be able to request IP Addresses from a defined pool of addresses. IPAM providers should then be able to find and fulfil such requests using any custom logic to do so.
+
+The contract consists of three resources:
+
+- An IPAddressClaim, which is used to allocate addresses
+- An IPAddress that IPAM providers use to fulfil IPAddressClaims
+- IP Pools, which hold provider specific configuration on how addresses should be allocated, e.g. a specific subnet to use
+
+Both **IPAddressClaims** and **IPAddresses** should be part of Cluster API, while **IPPools** are defined by the different IPAM providers.
+
+An **IPAddressClaim** is used by infrastructure providers to request an IP Address. The claim contains a reference to an IP Pool. Because the pool is provider specific, the IPAM controllers can decide whether to handle a claim by inspecting the group and kind of the pool reference.
+
+If a IPAM controller detects a Claim that references a Pool it controls, it allocates an IP address from that pool and creates an **IPAddress** to fulfil the claim. It also updates the status of the IPAddressClaim with a reference to the created Address.
+
+
+### Pools & IPAM Providers
+
+In order for IPAM providers to fulfil IPAddressClaims, they likely require a few parameters, e.g. the network from which to allocate an IP address. If there are multiple different IPAM providers, they also need to know which claims they are supposed to fulfil, and which to ignore.
+
+Therefore, IPAM providers are required to bring a custom IP Pool resource. The resource can have an arbitrary structure with the specific parameters for the IPAM system in use, and by inspecting the API group and kind of the pool reference held by IPAddressClaims, the provider can determine whether it is supposed to handle the claim or not.
+
+When an IPAM provider finds a IPAddressClaim that references one of its Pools, it allocates an IP address, creates an IPAddress resource and updates the status of the IPAddressClaim with a reference to this IPAddress.
+
+#### Examples
+
+A simple example for an IP Pool could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InClusterIPPool
```

Comment:
Saw the corresponding comments in Additional Notes. Might absolutely make sense to implement it in a separate repo, but if we end up implementing significant code in the core repo, I think it would be good to have a reference implementation in the core repo at some point to make testing easier. (like with CAPD)

Reply:
I have a reference implementation already, just waiting on a repo to share it. The proposal also mentions this in the implementation section. That reference implementation should also contain shared libraries that can be used by providers (both infra and ipam).

Reply:
Got it.

Depending on with which solution we end up, it sounds like we might end up with dependencies between:
* Core <= reference impl / lib
* Core <= infra providers
* reference impl / lib <= infra providers

Which is fine as it's not a circular dependency, but it means that the libraries have to upgrade to the latest Core CAPI version before infra provider can pick that up.

I guess that probably means if the reference impl is in a separate repo and the types are in core CAPI we should maybe add the library to core CAPI instead.

Reply:
I'd prefer to add the library to core. It was just my impression that the preference would be to keep this separate, as it doesn't concern all providers. On the other hand that shared library probably isn't going to be that big. Just some helpers to create addresses, and query claims and addresses I'd imagine.

Reply:
Maybe we can just infer from whatever we decide to do regarding maybe adding a field for ipPool to a core type